src/handler/core.rs

```diff
@@ -44,36 +52,71 @@ use crate::{description, HandlerDescription};
 /// This is very crucial when `b` is a filter: if it is chained, it decides
 /// whether or not to call `c`, but when it is branched, whether `c` is called
 /// depends solely on `a`.
-pub struct Handler<'a, Input, Output, Descr = description::Unspecified> {
-    data: Arc<HandlerData<Descr, DynF<'a, Input, Output>>>,
+pub struct Handler<'a, Output, Descr = description::Unspecified> {
+    data: Arc<HandlerData<Descr, DynF<'a, Output>>>,
 }
 
 struct HandlerData<Descr, F: ?Sized> {
     description: Descr,
+    sig: HandlerSignature,
     f: F,
 }
 
-type DynF<'a, Input, Output> =
-    dyn Fn(Input, Cont<'a, Input, Output>) -> HandlerResult<'a, Input, Output> + Send + Sync + 'a;
+/// A run-time handler signature. Used only for run-time type inference and
+/// checking of handler chains.
+///
+/// See [`crate::type_check`].
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum HandlerSignature {
+    /// The signature corresponding to [`crate::entry`].
+    Entry,
+    /// The handler signature with exact input and output types.
+    Other {
+        /// The set of types that the handler accepts. In the case of a DI
+        /// container, this is the set of types that this handler retrieves from
+        /// the container.
+        input_types: HashSet<Type>,
+        /// The set of types that this handler passes down the chain.
+        output_types: HashSet<Type>,
+        /// The set of "obligations", i.e., code locations where specific input
+        /// types are required.
+        obligations: HashMap<Type, &'static Location<'static>>,
+    },
+}
+
+/// A run-time representation of a type. Used only for run-time type inference
+/// and checking of handler chains.
+///
+/// See [`crate::type_check`].
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
```

Comment:
`Hash` impl should only really consider `id`, there is no reason to hash `name`, since it is always the same for a given `id`. `PartialEq` impl should also probably have a fast path for `ptr::eq(a.name, b.name)`.

Reply:
(ideally we'd only have `TypeId` as a key, and had name in the _value_, but if it's simpler to have `Type` as a key, let's make it a bit nicer)

Reply:
I don't understand why would we need `ptr::eq` on type names instead of just comparing type IDs, which are just two machine words in size. Otherwise, done.